Fix setting color + brightness of Tuya lights

BNolet opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Changing brightness alone does not change work mode, but changing brightness with a color value will keep the light in white mode. 

By verifying the new state has color or not, rather than the existing state being in color work mode, the light will change to color correctly by not explicitly setting the brightness.

Tuya interprets HSV as including the brightness in the (v) value (which is generally what that's used for when setting HSV values). The brightness value given by Home Assistant is still used in this case.
## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: #82759 #59056 #64142
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

BNolet committed:
Check if changing TO a color mode

Changing brightness alone does not change work mode, but changing brightness with a color value will keep the light in white mode. 

By verifying the new state has color or not, rather than the existing state being in color work mode, the light will change to color correctly.

Tuya interprets HSV as including the brightness in the (v) value (which is generally what that's used for when setting HSV values). The brightness value given by Home Assistant is still used in this case.

frenck commented on homeassistant/components/tuya/light.py:
```diff
@@ -542,7 +542,7 @@ def turn_on(self, **kwargs: Any) -> None:
         if (
             ATTR_BRIGHTNESS in kwargs
-            and self.color_mode != ColorMode.HS
+            and ATTR_HS_COLOR not in kwargs
```
This change is not correct. In case a service call is made that just changes brightness (thus no `ATTR_HS_COLOR` in the kwargs), while still the hs color mode isn't handled by this call (as brightness is part of the color).

../Frenck

BNolet replied:
I will perform some further testing and note what I tested but I'm pretty sure that I tested the scenario where only the brightness was changed and there was no issue caused concerning erroneous changing of the work mode. 

I can perform some further testing and note the combinations I perform, and report back here. However, if there's a better way to show my testing, I'd be happy to go through the steps required to perform the more in-depth methods of testing.

BNolet replied:
Okay so it turns out I did not test the actual use case that I saw the original code checking for in the brightness section: Changing only the brightness while in colour work mode. 

Instantly set the work mode back to white like ya'll predicted. My apologies for the mistake.  I like @emontnemery's solution so I will give that a shot

BNolet replied:
[9177c61](https://github.com/home-assistant/core/pull/88470/commits/9177c61147cfebf4bccd67b1b2ee9bac8fb61b3e) resolves the issue + made sure to note my tests. The checkmark indicates that the tests behaved as expected.

BNolet committed:
Merge branch 'home-assistant:dev' into patch-1

BNolet committed:
Fix brightness-only turning colour mode to white

This will take into account the case where brightness is the only parameter for both the case of colour mode and white mode.

Tests passed after this change:

* Brightness only (colour mode) ✅
* Brightness only (white mode) ✅
* Colour only (colour mode) ✅
* Colour only (white mode) ✅
* Colour temp only (colour mode) ✅
* Colour temp only (white mode) ✅
* Colour + brightness (colour mode) ✅
* Colour + brightness (white mode) ✅
* Colour temp + brightness (colour mode) ✅
* Colour temp + brightness (white mode) ✅

BNolet committed:
Merge branch 'dev' into patch-1

BNolet committed:
Merge branch 'dev' into patch-1

frenck reviewed (changes requested):
Some jobs in our CI are failing. Please check the output of those jobs to see what went wrong and adjust the PR accordingly.

Please un-draft it once the CI passes and it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

BNolet committed:
Merge branch 'home-assistant:dev' into patch-1

BNolet committed:
Fix code formatting

BNolet committed:
Merge branch 'dev' into patch-1

BNolet commented:
I believe this should fix the formatting issues with the CI

BNolet committed:
Merge branch 'dev' into patch-1

emontnemery reviewed (approved):
LGTM, thanks @BNolet 👍

frenck reviewed (approved):
Tested locally, and seems to work as expected.

Thanks @BNolet  👍
